### PR TITLE
HIVE-27867: Incremental materialized view throws NPE whew Iceberg source table is empty

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
* Use Objects.equals for comparing saved snapshot in MV metadata and table current snapshot.
* HiveStorageHandler.getCurrentSnapshotContext() returns null when table is empty and no snapshot exists. Add null check of snapshotId when creating filter predicate

### Why are the changes needed?
If both saved snapshot in MV metadata and table current snapshot are null NPE was thrown.

### Does this PR introduce _any_ user-facing change?
No.

### Is the change a dependency upgrade?
No.

### How was this patch tested?
```
mvn test -Dtest=TestHiveAugmentSnapshotMaterializationRule,TestHivePushdownSnapshotFilterRule -pl ql
mvn test -Dtest.output.overwrite -Dtest=TestIcebergLlapLocalCliDriver -Dqfile=mv_iceberg_orc.q -pl itests/qtest-iceberg -Piceberg -Pitests
```
and search for NPE in `hive.log`

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java [hunk 1]
```diff
@@ -125,11 +131,12 @@ public void onMatch(RelOptRuleCall call) {
     Table table = hiveTable.getHiveTableMD();
 
     SnapshotContext mvMetaTableSnapshot = mvMetaStoredSnapshot.get(table.getFullyQualifiedName());
-    if (mvMetaTableSnapshot.equals(table.getStorageHandler().getCurrentSnapshotContext(table))) {
+    if (Objects.equals(mvMetaTableSnapshot, table.getStorageHandler().getCurrentSnapshotContext(table))) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:134]: is there a null check for `table.getStorageHandler()`?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java [hunk 2]
```diff
@@ -124,7 +126,8 @@ private boolean setSnapShotId(RexNode op1, RexNode op2) {
       }
 
       RelOptHiveTable hiveTable = (RelOptHiveTable) relOptTable;
-      hiveTable.getHiveTableMD().setVersionIntervalFrom(Long.toString(snapshotId));
+      String snapshotIdText = snapshotId != NULL_SNAPSHOT_ID ? Long.toString(snapshotId) : null;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:129]: why is this needed if eventually, you just map null-> NULL_SNAPSHOT_ID -> null ?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:139]: Added null check.
In theory this can not happen because this rule is used only if MV base tables are Iceberg. Mixed mode is not supported.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:129]: got it, thanks! won't it work if we have `NOT(ISNULL($3)) AND <=($3, NULL)`? if not, that's OK with me

Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:129]: it is still `null`:
```
<=($3, NULL) -> null
anything AND null -> null
```

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:129]: so there is no lazy evaluation for AND operators? if left is false -> return false?

**Review by deniskuzZ — APPROVED** [review]
LGTM +1, since AND(FALSE, NULL) is not lazy evaluated to FALSE in Hive

**Comment by kasakrisz** [comment]
@deniskuzZ 
I revisited your proposal and found that `NOT(ISNULL($3)) AND <=($3, NULL)` is actually always simplified to `<=($3, NULL)` at Filter operator creation becasue $3 is a referenc to SnapshotID virtual column which type is `bigint not null`. I also checked that we don't need an incremental MV rebuild when any of the base tables was empty when the MV was last rebuilt because when the MV definition has 
* only one base table and that has new records inserted it actually means that all records are new and a full scan is needed.
* join. It is always inner join (only this join type is supported by the rewrite algorithm) hence we need full scan on all tables and a full rebuild because those new records must be joined with the records coming from the other table.

mv_iceberg_orc.q tests this scenario. In this case the augmented plan has a filter with a predicate `<=($3, NULL)` on top of the previously empty table and the rewrite doesn't happen and we fall back to full rebuild which is expected.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java [hunk 3]
```diff
@@ -125,11 +129,20 @@ public void onMatch(RelOptRuleCall call) {
...
+          "Mixing native and non-native tables in a materialized view definition is currently not supported!",
+          table.getFullyQualifiedName()));
+    }
+    if (Objects.equals(mvMetaTableSnapshot, table.getStorageHandler().getCurrentSnapshotContext(table))) {
       return;
     }
 
-    table.setVersionIntervalFrom(Long.toString(mvMetaTableSnapshot.getSnapshotId()));
+    Long snapshotId = null;
+    if (mvMetaTableSnapshot != null) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:142]:
````
Long snapshotId = mvMetaTableSnapshot != null ? mvMetaTableSnapshot.getSnapshotId() : null;
table.setVersionIntervalFrom(String.valueOf(snapshotId));
````

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java [hunk 4]
```diff
@@ -116,15 +116,16 @@ private boolean setSnapShotId(RexNode op1, RexNode op2) {
...
+      Long snapshotId = literal.getValueAs(Long.class);
 
       RelOptTable relOptTable = getRelOptTableOf(op2);
       if (relOptTable == null) {
         return false;
       }
 
       RelOptHiveTable hiveTable = (RelOptHiveTable) relOptTable;
-      hiveTable.getHiveTableMD().setVersionIntervalFrom(Long.toString(snapshotId));
+      String snapshotIdText = snapshotId != null ? Long.toString(snapshotId) : null;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:127]:
````
hiveTable.getHiveTableMD().setVersionIntervalFrom(String.valueOf(snapshotId))
````

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java [hunk 5]
```diff
@@ -141,11 +154,8 @@ public void onMatch(RelOptRuleCall call) {
     relBuilder.push(tableScan);
     List<RexNode> conds = new ArrayList<>();
     final RexNode literalHighWatermark = rexBuilder.makeLiteral(
-        mvMetaTableSnapshot.getSnapshotId(), snapshotIdType(relBuilder), false);
-    conds.add(
-        rexBuilder.makeCall(
-            SqlStdOperatorTable.LESS_THAN_OR_EQUAL,
-            ImmutableList.of(snapshotIdInputRef, literalHighWatermark)));
+        snapshotId, snapshotIdType(relBuilder.getTypeFactory()), false);
+    conds.add(rexBuilder.makeCall(SqlStdOperatorTable.LESS_THAN_OR_EQUAL, snapshotIdInputRef, literalHighWatermark));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:158]: original code formatting (with new lines) was more readable

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 3 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:142]: `String.valueOf(null)` returns string literal `"null"`. This is not a valid snapshotId

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java (hunk 4 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:127]: `String.valueOf(null)` returns string literal `"null"`. This is not a valid snapshotId

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 5 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:158]: Sorry, where were the new lines?

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:158]:
````
    conds.add(
        rexBuilder.makeCall(
            SqlStdOperatorTable.LESS_THAN_OR_EQUAL,
            ImmutableList.of(snapshotIdInputRef, literalHighWatermark)));
````

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 3 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:142]: sorry, `Objects.toString(obj, null)`

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java (hunk 4 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java:127]: Objects.toString(obj, null)

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 3 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java]: refactored

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java (hunk 4 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HivePushdownSnapshotFilterRule.java]: refactored

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java (hunk 5 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java]: refactored

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java [hunk 6]
```diff
@@ -139,14 +148,11 @@ public void onMatch(RelOptRuleCall call) {
...
     relBuilder.push(tableScan);
-    List<RexNode> conds = new ArrayList<>();
-    final RexNode literalHighWatermark = rexBuilder.makeLiteral(
-        mvMetaTableSnapshot.getSnapshotId(), snapshotIdType(relBuilder), false);
-    conds.add(
-        rexBuilder.makeCall(
-            SqlStdOperatorTable.LESS_THAN_OR_EQUAL,
-            ImmutableList.of(snapshotIdInputRef, literalHighWatermark)));
-    relBuilder.filter(conds);
+    final RexNode literalSnapshotId = rexBuilder.makeLiteral(
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java:151]: since you'll have to re-trigger build again, should we rename to `snapshotIdLiteral`

Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAugmentSnapshotMaterializationRule.java]: renamed
